[SPARK-7347] Dag visualization: add hover to RDDs on job page

andrewor14 opened the pull request:
Add tooltips to the dots on the job page so it's clearer what they represent.

_By the way, most of the commits are from a left-over PR that's already merged. Really, only one of these commits actually belong to this PR, and this has caused me many unnecessary merge conflicts..._

<img src="https://issues.apache.org/jira/secure/attachment/12730538/job-page-hover.png" width="300px"></img>

Commit:
Scope all RDD methods

This commit provides a mechanism to set and unset the call scope
around each RDD operation defined in RDD.scala. This is useful
for tagging an RDD with the scope in which it is created. This
will be extended to similar methods in SparkContext.scala and
other relevant files in a future commit.

Commit:
Add a few missing scopes to certain RDD methods

Commit:
Expose the necessary information in RDDInfo

This includes the scope field that we added in previous commits,
and the parent IDs for tracking the lineage through the listener
API.

Commit:
Translate RDD information to dot file

It turns out that the previous scope information is insufficient
for producing a valid dot file. In particular, the scope hierarchy
was missing, but crucial to differentiate between a parent RDD
being in the same encompassing scope and it being in a completely
distinct scope. Also, unique scope identifiers are needed to
simplify the code significantly.

This commit further adds the translation logic in a UI listener
that converts RDDInfos to dot files.

Commit:
First working implementation of visualization with vis.js

Commit:
Re-implement scopes through annotations instead

The previous "working" implementation frequently ran into
NotSerializableExceptions. Why? ClosureCleaner doesn't like
closures being wrapped in other closures, and these closures
are simply not cleaned (details are intentionally omitted here).

This commit reimplements scoping through annotations. All methods
that should be scoped are now annotated with @RDDScope. Then, on
creation, each RDD derives its scope from the stack trace, similar
to how it derives its call site. This is the cleanest approach
that bypasses NotSerializableExceptions with least significant
limitations.

Commit:
Revert a few unintended style changes

Commit:
Move RDD scope util methods and logic to its own file

Just a small code re-organization.

Commit:
Merge branch 'master' of github.com:apache/spark into viz

Commit:
Reimplement rendering with dagre-d3 instead of viz.js

Before this commit, this patch relies on a JavaScript version of
GraphViz that was compiled from C. Even the minified version of
this resource was ~2.5M. The main motivation for switching away
from this library, however, is that this is a complete black box
of which we have absolutely no control. It is not at all extensible,
and if something breaks we will have a hard time understanding
why.

The new library, dagre-d3, is not perfect either. It does not
officially support clustering of nodes; for certain large graphs,
the clusters will have a lot of unnecessary whitespace. A few in
the dagre-d3 community are looking into a solution, but until then
we will have to live with this (minor) inconvenience.

Commit:
Merge branch 'master' of github.com:apache/spark into viz

Commit:
Fill in documentation + miscellaneous minor changes

For instance, this adds ability to throw away old stage graphs.

Commit:
Embed the viz in the UI in a toggleable manner

Commit:
Re-implement scopes using closures instead of annotations

The problem with annotations is that there is no way to associate
an RDD's scope with another's. This is because the stack trace
simply does not expose enough information for us to associate one
instance of a method invocation with another.

So, we're back to closures. Note that this still suffers from the
same not serializable issue previously discussed, and this is being
fixed in the ClosureCleaner separately.

Commit:
Ensure that HadoopRDD is actually serializable

Commit:
Add ID to node label (minor)

Commit:
Rat excludes

Commit:
Merge branch 'master' of github.com:apache/spark into viz2

Commit:
Fix scala style

Commit:
Merge branch 'master' of github.com:apache/spark into viz2

Commit:
Remove all return statements in `withScope`

The closure cleaner doesn't like these statements, for a good
reason.

Commit:
Add parent IDs to StageInfo

Commit:
Fix line too long

Commit:
Remove more return statements from scopes

Commit:
Introduce visualization to the Job Page

This includes a generalization of the visualization previously
displayed on the stage page. More functionality is needed in
JavaScript to prevent the job visualization from looking too
cluttered. This is still WIP.

Commit:
Connect RDDs across stages + update style

This requires us to track incoming and outgoing edges in each
stage on the backend, and render the connecting edges manually
ourselves in d3.

Commit:
Translate stage boxes taking into account the width

Previously we had a lot of overlapping boxes for say ALS. This is
because we did not take into account of the widths of the previous
boxes.

Commit:
Add link from jobs to stages

Commit:
Merge branch 'master' of github.com:apache/spark into viz2

Conflicts:
	core/src/main/scala/org/apache/spark/storage/RDDInfo.scala
	core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala
	core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala

Commit:
Refactor + clean up + document JS visualization code

This commit should not introduce any substantial functionality
differences. It just cleans up the JavaScript side of this patch
such that it is easier to follow.

Commit:
Separate HTML generating code from listener

Commit:
Rename references to viz scopes to viz clusters

The term 'scope' is really overloaded. A scope should be logically
agnostic to how it is represented in the UI. We rename all such
references to 'cluster', which represents either an operator scope,
a stage, or a job.

Commit:
Implement a programmatic representation of operator scopes

Rather than doing our own string parsing magic, serialize the scopes
to and from JSON and operate directly on the class that implements
these scopes.

Commit:
Extract visualization logic from listener

Commit:
Rename OperatorScope -> RDDOperationScope

Commit:
Display cached RDDs on the viz

Commit:
Merge branch 'master' of github.com:apache/spark into viz2

Commit:
Fix tests

Commit:
Fix python tests

Commit:
Merge branch 'master' of github.com:apache/spark into viz2

Commit:
Fix python style

Commit:
children -> child (minor)

Commit:
Restore "null" as default for RDD name

Instead, RDDInfo's name field will default to the RDD class name
if the RDD's name is not defined. This avoids changing RDD#toString
behavior.

Commit:
Wrap a few more RDD functions in an operation scope

Commit:
Add tests for RDDOperationScope

Commit:
Change RDD cache color to red (minor)

Commit:
Round corners of RDD boxes on stage page (minor)

Commit:
Add tooltips to RDDs on job page

Commit:
Merge branch 'master' of github.com:apache/spark into viz-hover

Conflicts:
	core/src/main/resources/org/apache/spark/ui/static/spark-dag-viz.js
	core/src/main/scala/org/apache/spark/rdd/RDD.scala
	core/src/main/scala/org/apache/spark/ui/SparkUI.scala
	core/src/main/scala/org/apache/spark/ui/UIUtils.scala
	core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala
	core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala
	core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31888 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31888/consoleFull) for   PR 5912 at commit [`516c930`](https://github.com/apache/spark/commit/516c930b0d3976b2c94c7e823a417747842ed746).

SparkQA commented:
[Test build #31888 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31888/consoleFull) for   PR 5912 at commit [`516c930`](https://github.com/apache/spark/commit/516c930b0d3976b2c94c7e823a417747842ed746).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class EnumUtil `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31888/
Test FAILed.

Commit:
Fix RAT

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31895 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31895/consoleFull) for   PR 5912 at commit [`07f25c3`](https://github.com/apache/spark/commit/07f25c3b08a99f39290e84ea80d630c369d9d55a).

SparkQA commented:
[Test build #31895 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31895/consoleFull) for   PR 5912 at commit [`07f25c3`](https://github.com/apache/spark/commit/07f25c3b08a99f39290e84ea80d630c369d9d55a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class EnumUtil `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31895/
Test PASSed.

Commit:
Merge branch 'master' of github.com:apache/spark into viz-hover

Conflicts:
	core/src/main/resources/org/apache/spark/ui/static/spark-dag-viz.js

Commit:
Merge branch 'master' of github.com:apache/spark into viz-hover

Commit:
Fix CSS class selection issue

The new div is not actually in the SVG itself, so we need to use
a more general selector.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32056 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32056/consoleFull) for   PR 5912 at commit [`4fb4545`](https://github.com/apache/spark/commit/4fb4545d3faa4a674f491c689bb83330a67a79cf).

andrewor14 commented:
Closing in favor of #5957

SparkQA commented:
[Test build #32056 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32056/consoleFull) for   PR 5912 at commit [`4fb4545`](https://github.com/apache/spark/commit/4fb4545d3faa4a674f491c689bb83330a67a79cf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32056/
Test PASSed.
